### PR TITLE
Initial weights compresson for pytorch models to 8 bit.

### PR DESCRIPTION
**Pull request by andreyanufr**
### Changes

Added data free int8 weight compression implementation for pytorch models.

### Reason for changes

Acceleration of compression and next conversion to IR

### CODE REVIEW

File: nncf/torch/quantization/quantize_model.py [hunk 1]
```diff
@@ -256,3 +258,13 @@ def send_to_device(tensor):
     compressed_model.nncf.disable_dynamic_graph_building()
 
     return compressed_model
+
+
+def weights_compression_impl(model: torch.nn.Module, compress_weights: bool = False) -> torch.nn.Module:
```
Comment (alexsu52) [inline nncf/torch/quantization/quantize_model.py:263]: Please, rename weights_compression_impl -> compress_weights

File: nncf/torch/quantization/weights_compression.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+    def __init__(self, zero_point, scale):
+        super().__init__()
+        self.zero_point = zero_point
+        self.scale = scale
+
+    def forward(self, input, op_arg):
+        if input.weight.dtype is torch.uint8:
+            w = input.weight.type(dtype=self.scale.dtype)
+            input.weight = (w - self.zero_point) * self.scale
```
Comment (ljaljushkin) [inline nncf/torch/quantization/weights_compression.py:28]: Is `input` actually layer? It looks confusing when weights are taken from input, usually weights are multiplied with input. IMO, `layer` or `module` is a better name. What do you think?

File: nncf/torch/quantization/weights_compression.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    module: nn.Module, allowed_types: Dict, compress_weights=False
+) -> Optional[nn.Module]:
+    """
+    Insets weights compression with dequantization or quantization pre operation for Linear and Embedding layers.
+
+    :param module: The module to insert the weights compression.
+    :param compress_weights: Enables real compression of weights in Linear and Embedding layers.
+        If False inserts pytorch torch.fake_quantize_per_channel_affine(),
+        else compress weights to int8 and inserts custom dequantization.
+    :param allowed_types: dist with pairs (allowed type for weights compression, dimention for scale and zero point)
```
Comment (ljaljushkin) [inline nncf/torch/quantization/weights_compression.py:46]: It's not just types. I think there can be a better name that would say something about compression dimension also. 
Maybe you even don't need this dict, just list of types, because `NNCFEmbedding` and `NNCFLinear` have this information in `target_weight_dim_for_compression`. Why not reuse this?

File: nncf/torch/quantization/weights_compression.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        If False inserts pytorch torch.fake_quantize_per_channel_affine(),
+        else compress weights to int8 and inserts custom dequantization.
+    :return: The module with inserted operations. The module is not trainable if compress_weights is True.
+    """
+    user_types = list(NNCF_WRAPPED_USER_MODULES_DICT.values())
+    allowed_types = {NNCFEmbedding: 0, NNCFLinear: 1}
+
+    for user_type in user_types:
+        if torch.nn.Embedding in user_type.__mro__:
+            allowed_types[user_type] = 0
```
Comment (ljaljushkin) [inline nncf/torch/quantization/weights_compression.py:88]: `get_target_dim_for_compression` should be 0 for `NNCF_WRAPPED_USER_MODULES_DICT`, by default. 
should be possible to reuse it

**Review by ljaljushkin — CHANGES_REQUESTED** [review]

File: nncf/torch/quantization/weights_compression.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+            input.weight = torch.fake_quantize_per_channel_affine(
+                input.weight, self.scale, self.zero_point, axis, 0, 255
+            )
+
+
+def _insert_pre_compression_operations(
+    module: nn.Module, allowed_types: Dict, compress_weights=False
+) -> Optional[nn.Module]:
+    """
+    Insets weights compression with dequantization or quantization pre operation for Linear and Embedding layers.
```
Comment (ljaljushkin) [inline nncf/torch/quantization/weights_compression.py:40]:
```suggestion
    Inserts weights compression with dequantization or quantization pre operation for Linear and Embedding layers.
```

File: nncf/torch/quantization/weights_compression.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+from torch import nn
+
+from nncf.torch.layers import NNCF_WRAPPED_USER_MODULES_DICT
+from nncf.torch.layers import NNCFEmbedding
+from nncf.torch.layers import NNCFLinear
+from nncf.torch.quantization.quantize_functions import get_scale_zp_from_input_low_input_high
+
+
+class WeightsCompressor(nn.Module):
+    """Class provided fake quantize operation or dequantization of compressed weights in forward pass
```
Comment (ljaljushkin) [inline nncf/torch/quantization/weights_compression.py:13]: This class tries to handle 2 responsibilities.  
IMO, there will be more clear and robust code with 2 separate classes for a certain task. 
First - Dequantizer
Second - TorchFakeQuantizer

File: nncf/torch/quantization/quantize_model.py [hunk 7]
```diff
@@ -256,3 +258,13 @@ def send_to_device(tensor):
     compressed_model.nncf.disable_dynamic_graph_building()
 
     return compressed_model
+
+
+def compress_weights(model: torch.nn.Module, compress_weights: bool = False) -> torch.nn.Module:
```
Comment (ljaljushkin) [inline nncf/torch/quantization/quantize_model.py:263]: Also, I found it confusing that `compress_weights` function can be launched with option `compress_weights=False`.
Seems like nothing will happen on that call.
I'd use something like this: `fake_quantize` or `save_int_data`.

File: nncf/torch/quantization/weights_compression.py [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+from nncf.torch.layers import NNCF_WRAPPED_USER_MODULES_DICT
+from nncf.torch.layers import NNCFEmbedding
+from nncf.torch.layers import NNCFLinear
+from nncf.torch.quantization.quantize_functions import get_scale_zp_from_input_low_input_high
+
+
+class WeightsDecompressor(nn.Module):
+    """Class decompression of compressed weights in forward pass
+
+    Atributes:
```
Comment (KodiaqQ) [inline nncf/torch/quantization/weights_compression.py:26]:
```suggestion
    Attributes:
```

File: nncf/torch/quantization/weights_compression.py [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+
+    def forward(self, layer, op_arg):
+        w = layer.weight.type(dtype=self.scale.dtype)
+        layer.weight = (w - self.zero_point) * self.scale
+
+
+class WeightsFQ(nn.Module):
+    """Class provided fake quantize operation in forward pass
+
+    Atributes:
```
Comment (KodiaqQ) [inline nncf/torch/quantization/weights_compression.py:44]:
```suggestion
    Attributes:
```

File: nncf/torch/quantization/weights_compression.py [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+        else compress weights to int8 and inserts custom dequantization.
+    :param allowed_types: list of allowed types for weights compression
+    :param lewel_high: highest  possible value of compressed weights (lower is 0 in assymetric quantization)
+    :return: The module with inserted operations. The module is not trainable if compress_weights is True.
+    """
+    for _, layer in module.named_children():
+        if not type(layer) in allowed_types:
+            _insert_pre_compression_operations(layer, allowed_types, compress_weights, lewel_high)
+            continue
+        q_dim = layer.target_weight_dim_for_compression
```
Comment (KodiaqQ) [inline nncf/torch/quantization/weights_compression.py:80]: I'm not sure that these variables are clearly explainable. Can we rename them?

File: nncf/torch/quantization/weights_compression.py [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+        self.axis = axis
+
+    def forward(self, layer, op_arg):
+        layer.weight = torch.fake_quantize_per_channel_affine(
+            layer.weight, self.scale, self.zero_point, self.axis, 0, 255
+        )
+
+
+def _insert_pre_compression_operations(
+    module: nn.Module, allowed_types: Dict, compress_weights=False, lewel_high=255
```
Comment (KodiaqQ) [inline nncf/torch/quantization/weights_compression.py:63]:
```suggestion
    module: nn.Module, allowed_types: Dict, compress_weights=False, level_high=255
```

**Review by KodiaqQ — COMMENTED** [review]
General comment: need tests.

File: nncf/torch/quantization/weights_compression.py (hunk 2 above)
Comment (andreyanufr) [reply nncf/torch/quantization/weights_compression.py]: Done

File: nncf/torch/quantization/weights_compression.py [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+        super().__init__()
+        self.zero_point = zero_point
+        self.scale = scale
+
+    def forward(self, input, op_arg):
+        if input.weight.dtype is torch.uint8:
+            w = input.weight.type(dtype=self.scale.dtype)
+            input.weight = (w - self.zero_point) * self.scale
+        elif not self.zero_point.dtype is torch.uint8:
+            axis = 0 if input.weight.shape[0] == self.scale.shape[0] else 1
```
[the remark this answers is not in the document]
Comment (andreyanufr) [reply nncf/torch/quantization/weights_compression.py]: Done

File: nncf/torch/quantization/weights_compression.py (hunk 4 above)
Comment (andreyanufr) [reply nncf/torch/quantization/weights_compression.py]: Done

File: nncf/torch/quantization/weights_compression.py (hunk 5 above)
Comment (andreyanufr) [reply nncf/torch/quantization/weights_compression.py:66]: Done

File: nncf/torch/quantization/quantize_model.py (hunk 7 above)
Comment (andreyanufr) [reply nncf/torch/quantization/quantize_model.py]: Done

File: nncf/torch/quantization/weights_compression.py (hunk 2 above)
Comment (andreyanufr) [reply nncf/torch/quantization/weights_compression.py]: Done

File: nncf/torch/quantization/weights_compression.py (hunk 3 above)
Comment (andreyanufr) [reply nncf/torch/quantization/weights_compression.py]: Done

File: nncf/torch/quantization/weights_compression.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (andreyanufr) [reply nncf/torch/quantization/weights_compression.py]: Done

File: nncf/torch/quantization/weights_compression.py (hunk 4 above)
Comment (andreyanufr) [reply nncf/torch/quantization/weights_compression.py]: Done

File: nncf/torch/quantization/quantize_model.py (hunk 1 above)
Comment (andreyanufr) [reply nncf/torch/quantization/quantize_model.py]: Done

File: nncf/torch/quantization/weights_compression.py (hunk 10 above)
Comment (andreyanufr) [reply nncf/torch/quantization/weights_compression.py]: Done

File: nncf/torch/quantization/weights_compression.py [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+        else compress weights to int8 and inserts custom dequantization.
+    :param bits: number of bits for compression/quantization. Note: compressed weights type is
+        uint8 with one element per 8 bit.
+    :return: The module with inserted operations. The module is not trainable if use_fake_qunatize is False.
+    """
+    user_types = list(NNCF_WRAPPED_USER_MODULES_DICT.values())
+    allowed_types = [NNCFEmbedding, NNCFLinear]
+    lewel_high = 2**bits - 1
+
+    assert lewel_high < 256
```
Comment (ljaljushkin) [inline nncf/torch/quantization/weights_compression.py:117]:
```suggestion
    assert level_high < 256
```

File: nncf/torch/quantization/weights_compression.py [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+        self.zero_point = zero_point
+        self.scale = scale
+
+    def forward(self, layer, op_arg):
+        w = layer.weight.type(dtype=self.scale.dtype)
+        layer.weight = (w - self.zero_point) * self.scale
+
+
+class WeightsFQ(nn.Module):
+    """Class provided fake quantize operation in forward pass
```
Comment (ljaljushkin) [inline nncf/torch/quantization/weights_compression.py:42]: Usually we start from verb omitting unnecessary `class`.
I'd write: `Replaces weights with Torch's FakeQuantize operation on forward pass`

File: nncf/quantization/quantize_model.py [hunk 15]
```diff
@@ -218,3 +218,26 @@ def quantize_with_accuracy_control(
...
     raise RuntimeError(f"Unsupported type of backend: {backend}")
+
+
+@api(canonical_alias="nncf.weights_compression")
+def weights_compression(model: TModel, use_fake_quantize: bool = False) -> TModel:
+    """
+    Applies per-channel weights compression with dequantization or fake quantize insertion.
+
+    :param model: A model to be compressed.
+    :type  model: TModel
```
Comment (ljaljushkin) [inline nncf/quantization/quantize_model.py:229]: Our code style guide doesn't require specifying parameter type.
Please remove to make it shorter and more consistent with the current code style.

File: nncf/torch/layer_utils.py [hunk 16]
```diff
@@ -24,6 +24,10 @@ def __init__(self, module):
     def __getattr__(self, name):
         return getattr(self._module, name)
 
+    @property
+    def __class__(self):
```
Comment (ljaljushkin) [inline nncf/torch/layer_utils.py:28]: Should it be part of another PR?

**Review by ljaljushkin — CHANGES_REQUESTED** [review]

File: nncf/quantization/quantize_model.py (hunk 15 above)
Comment (andreyanufr) [reply nncf/quantization/quantize_model.py]: Done

File: nncf/torch/quantization/weights_compression.py [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+from torch import nn
+
+from nncf.torch.layers import NNCF_WRAPPED_USER_MODULES_DICT
+from nncf.torch.layers import NNCFEmbedding
+from nncf.torch.layers import NNCFLinear
+from nncf.torch.quantization.quantize_functions import get_scale_zp_from_input_low_input_high
+
+
+class WeightsDecompressor(nn.Module):
+    """Class decompression of compressed weights in forward pass
```
[the remark this answers is not in the document]
Comment (andreyanufr) [reply nncf/torch/quantization/weights_compression.py]: Done

File: nncf/torch/quantization/weights_compression.py (hunk 14 above)
Comment (andreyanufr) [reply nncf/torch/quantization/weights_compression.py]: Done

File: nncf/torch/quantization/weights_compression.py [hunk 18]
```diff
@@ -0,0 +1,123 @@
...
+
+    def __init__(self, zero_point, scale, axis=0):
+        super().__init__()
+        self.zero_point = zero_point
+        self.scale = scale
+        self.axis = axis
+
+    def forward(self, layer, op_arg):
+        layer.weight = torch.fake_quantize_per_channel_affine(
+            layer.weight, self.scale, self.zero_point, self.axis, 0, 255
```
[the remark this answers is not in the document]
Comment (andreyanufr) [reply nncf/torch/quantization/weights_compression.py]: Done

File: nncf/torch/quantization/weights_compression.py (hunk 13 above)
Comment (andreyanufr) [reply nncf/torch/quantization/weights_compression.py]: Done

File: nncf/quantization/quantize_model.py (hunk 15 above)
Comment (alexsu52) [reply nncf/quantization/quantize_model.py]: This is incorrect comment. API function requires type for params. Please, take a look at https://github.com/openvinotoolkit/nncf/blob/aff8237371baaca0df3ac7c71172425a844fb7c3/nncf/quantization/quantize_model.py#L29

File: nncf/quantization/quantize_model.py [hunk 19]
```diff
@@ -217,3 +217,24 @@ def quantize_with_accuracy_control(
         )
 
     raise RuntimeError(f"Unsupported type of backend: {backend}")
+
+
+@api(canonical_alias="nncf.weights_compression")
+def weights_compression(model: TModel, use_fake_quantize: bool = False) -> TModel:
```
Comment (alexsu52) [inline nncf/quantization/quantize_model.py:223]:
```suggestion
def compress_weights(model: TModel, use_fake_quantize: bool = False) -> TModel:
```

**Review by alexsu52 — APPROVED** [review]
LGTM

File: nncf/quantization/quantize_model.py (hunk 15 above)
Comment (ljaljushkin) [reply nncf/quantization/quantize_model.py]: ok, havent't noticed. Style guide should be updated then.

**Review by KodiaqQ — APPROVED** [review]

**Review by ljaljushkin — APPROVED** [review]

File: nncf/quantization/quantize_model.py (hunk 19 above)
Comment (andreyanufr) [reply nncf/quantization/quantize_model.py]: Done

File: nncf/quantization/quantize_model.py [hunk 20]
```diff
@@ -217,3 +217,24 @@ def quantize_with_accuracy_control(
...
+    """
+    Applies per-channel weights compression with dequantization or fake quantize insertion.
```
[the remark this answers is not in the document]
Comment (andreyanufr) [reply nncf/quantization/quantize_model.py]: Done

**Comment by andreyanufr** [comment]
jenkins please retry a build

**Comment by andreyanufr** [comment]
run pytorch pre-commit tests

**Comment by andreyanufr** [comment]
run pytorch pre-commit tests
